src/json.cpp

```diff
@@ -0,0 +1,190 @@
+                    if (name_sexp == R_NaString) {
+                        json_error(sexp, "all elements in list must be named, but [[%d]] is not", i + 1);
+                    }
+
+                    const char* name = R_CHAR(name_sexp);
+
+                    auto insert_result = result.insert(std::make_pair(name, js::value()));
+                    if (!insert_result.second) {
+                        json_error(sexp, "duplicate name '%s' in list", name);
+                    }
+
+                    auto iter = insert_result.first;
+                    auto& elem = iter->second;
+
+                    SEXP elem_sexp = VECTOR_ELT(sexp, i);
+                    if (!to_json(elem_sexp, elem)) {
+                        result.erase(iter);
+                    }
+                }
+            }
+
+            void env_to_object(SEXP sexp, js::object& result) {
+                SEXP names = R_lsInternal3(sexp, R_TRUE, R_FALSE);
+                R_len_t count = Rf_length(names);
+
+                for (R_len_t i = 0; i < count; ++i) {
+                    SEXP name_sexp = STRING_ELT(names, i);
+                    const char* name = R_CHAR(name_sexp);
+
+                    auto insert_result = result.insert(std::make_pair(name, js::value()));
+                    if (!insert_result.second) {
+                        json_error(sexp, "duplicate name '%s' in environment", name);
+                    }
+
+                    auto iter = insert_result.first;
+                    auto& elem = iter->second;
+
+                    SEXP sym_sexp = Rf_installChar(name_sexp);
+                    SEXP elem_sexp = Rf_findVar(sym_sexp, sexp);
+                    if (!to_json(elem_sexp, elem)) {
+                        result.erase(iter);
+                    }
+                }
+            }
+        }
+
+        bool to_json(SEXP sexp, js::value& result) {
+            if (Rf_length(sexp) == 0) {
+                result = js::value();
+                return true;
+            }
+
+            switch (TYPEOF(sexp)) {
+            case NILSXP:
+                result = js::value();
+                return true;
+
+            case VECSXP: {
+                SEXP names = Rf_getAttrib(sexp, R_NamesSymbol);
+                if (Rf_isNull(names)) {
+                    result = js::value(js::array());
+                    list_to_array(sexp, result.get<js::array>());
+                } else {
+                    result = js::value(js::object());
+                    list_to_object(sexp, names, result.get<js::object>());
+                }
+                return true;
+            }
+
+            case ENVSXP:
+                result = js::value(js::object());
+                env_to_object(sexp, result.get<js::object>());
+                return true;
+
+            case LGLSXP: {
+                at_most_one(sexp);
+                int x = *LOGICAL(sexp);
+                result = x == R_NaInt ? js::value() : js::value(x != 0);
```

Comment:
In the standard R headers, yes. There isn't one in ours precisely because it's just an alias, so you're just supposed to use `R_NaInt` directly in this case.